File: delightreading-server/src/main/java/com/delightreading/user/model/UserGroupEntity.java
package com.delightreading.user.model;

import com.delightreading.common.BaseEntity;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.time.Instant;
import java.util.LinkedHashSet;

@Entity
@Table(name = "user_group")

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class UserGroupEntity extends BaseEntity {

    public static final String STATUS_ACTIVE = "active";


    // family, academic, club, etc.
    @Column(name = "type")
    @Enumerated(EnumType.STRING)
    UserGroupType type;

    @Column(name = "name")
    String name;

    @Column(name = "description")
    String description;

    @Column(name = "picture_uri")
    String pictureUri;

    @Column(name = "cover_image_uri")
    String coverImageUri;

    @Column(name = "category")
    String category;

    @Column(name = "rules")
    String rules;

    @Column(name = "website")
    String website;

    @Column(name = "start_date")
    Instant startDate;

    @Column(name = "close_date")
    Instant closeDate;

    @Column(name = "group_status")
    String groupStatus;

    // Integer memberCount;

    @Transient
//    @OneToMany(
//            mappedBy = "user_group",
//            cascade = CascadeType.ALL,
//            orphanRemoval = true
//    )
    LinkedHashSet<UserGroupMemberEntity> members = new LinkedHashSet<>();

    public void addMember(UserGroupMemberEntity member) {
        members.add(member);
    }
}

File: delightreading-backend/src/main/kotlin/com/delightreading/user/model/UserGroupEntity.kt
package com.delightreading.user.model

import java.io.Serializable
import java.time.Instant
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "user_group")
data class UserGroupEntity(

    //<editor-fold desc="Common Entity Base">
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var sid: Long? = null,

    @Column(name = "uid")
    var uid: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    var status: UserAccountStatus = UserAccountStatus.ACTIVE,

    @Column(name = "created_by")
    var createdBy: String? = null,

    @Column(name = "created_at")
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_by")
    var updatedBy: String? = null,

    @Column(name = "updated_at")
    var updatedAt: Instant? = null,
    //</editor-fold>


    // family, academic, club, etc.
    @Column(name = "type")
    @Enumerated(EnumType.STRING)
    var type: UserGroupType? = null,

    @Column(name = "name")
    var name: String? = null,

    @Column(name = "description")
    var description: String? = null,

    @Column(name = "picture_uri")
    var pictureUri: String? = null,

    @Column(name = "cover_image_uri")
    var coverImageUri: String? = null,

    @Column(name = "category")
    var category: String? = null,

    @Column(name = "rules")
    var rules: String? = null,

    @Column(name = "website")
    var website: String? = null,

    @Column(name = "start_date")
    var startDate: Instant? = null,

    @Column(name = "close_date")
    var closeDate: Instant? = null,

    @Column(name = "group_status")
    var groupStatus: String? = null,

// Integer memberCount;

    @Transient
//    @OneToMany(
//            mappedBy = "user_group",
//            cascade = CascadeType.ALL,
//            orphanRemoval = true
//    )
    var members: LinkedHashSet<UserGroupMemberEntity> = LinkedHashSet<UserGroupMemberEntity>()//
) : Serializable {

    //<editor-fold desc="Common Entity Base">
    @PrePersist
    fun prePersist() {
        if (uid == null) {
            uid = UUID.randomUUID().toString()
        }
        createdAt = Instant.now()
        // createdBy = LoggedUser.get();
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = Instant.now()
        // updatedBy = LoggedUser.get();
    }
    //</editor-fold>

}